### PR TITLE
 Embrace the hellscripting

### PR DESCRIPTION
**Pull request by lugray**
Apparently aliasing reserved words is a thing that shells allow.  And
it's also apparently a thing that users do.

Challenge accepted.

Builtins can be escaped with `\`, but reserved words cannot, and must
simply be avoided.

Looping got reimplemented with recursion.

```sh
if cond; then
  some
  statements
else
  other
  things
fi
```

becomes

```sh
cond && {
  some
  statements
} || {
  other
  things
}
```

### CODE REVIEW

**Comment by lugray** [comment]
Fixes #8.

**Review by donk-shopify — APPROVED** [review]
My god it's full of stars.

**Review by lavoiesl — APPROVED** [review]
LGTM. I understand what this is going for, but I’m having a hard time reviewing such hellscripting.

File: hookbook.sh [hunk 1]
```diff
@@ -23,23 +23,26 @@ __hookbook_shell="$(\ps -p $$ | \awk 'NR > 1 { sub(/^-/, "", $4); print $4 }')"
...
-      in=0
-      break
-    fi
-  done
-  return $in
+  [[ $# -lt 2 ]] && \return 1
+  \local seeking=$1; \shift
+  \local check=$1; \shift
+  [[ $seeking == $check ]] && \return 0
+  __hookbook_array_contains $seeking "$@"
```
Comment (burke) [inline hookbook.sh]: trivial style point: write variables as `"${seeking}"`, except one-character automatic variables, which should be `"$1"`

File: hookbook.sh [hunk 2]
```diff
@@ -50,12 +53,17 @@ case "${__hookbook_shellname}" in
       __hookbook_array_contains "__hookbook_${fn}_precmd" "${precmd_functions[@]}" \
         || precmd_functions+=("__hookbook_${fn}_precmd")
     }
+}
 
-    ;;
-  bash)
-    if ! declare -p __hookbook_functions >/dev/null 2>&1; then
+__hookbook_setup_bash() {
+    # declare -p __hookbook_functions >/dev/null 2>&1 || __hookbook_functions=()
```
Comment (burke) [inline hookbook.sh]: commented code, probably best to remove.

File: hookbook.sh [hunk 3]
```diff
@@ -50,12 +53,17 @@ case "${__hookbook_shellname}" in
...
+    # ! declare -p __hookbook_functions >/dev/null 2>&1
+    declare -p __hookbook_functions >/dev/null 2>&1 || {
       __hookbook_functions=()
-    fi
+    }
+    # if ! declare -p __hookbook_functions >/dev/null 2>&1; then
```
Comment (burke) [inline hookbook.sh]: idem

File: hookbook.sh [hunk 4]
```diff
@@ -69,37 +77,31 @@ case "${__hookbook_shellname}" in
...
-      elif [[ \"\$(${__stat_stderr})\" == \"${__dev_null_major}\" ]]; then
-        return
-      fi
-      for fn in \"\${__hookbook_functions[@]}\"; do
-        \${fn} preexec
-      done
+    }
+    \eval "__hookbook_debug_handler() {
+      [[ \"\${BASH_COMMAND}\" == \"\${PROMPT_COMMAND}\" ]] && \\return
+      [[ \"\$(${__stat_stderr})\" == \"${__dev_null_major}\" ]] && \\return
```
Comment (burke) [inline hookbook.sh]: :joy: This line is pretty :fire:☠️ :fire:

File: hookbook.sh [hunk 5]
```diff
@@ -108,36 +110,35 @@ case "${__hookbook_shellname}" in
...
         #   PROMPT_COMMAND="${fn}; ${PROMPT_COMMAND}"
         # ...except with weird magic to toggle off `-x` if it's set, much like
         # in the DEBUG trap above.
         PROMPT_COMMAND="{
-          if [[ \$- =~ x ]];
-          then set +x; ${fn} precmd 2>&3; set -x;
-          else ${fn} precmd 2>&3;
-          fi;
+          [[ \$- =~ x ]] && {
+            set +x; ${fn} precmd 2>&3; set -x;
```
Comment (burke) [inline hookbook.sh]: not as many `\` here. should there be?

File: hookbook.sh [hunk 6]
```diff
@@ -108,36 +110,35 @@ case "${__hookbook_shellname}" in
...
-    >&2 \echo "hookbook is not compatible with your shell (${__hookbook_shell})"
-    \return 1
-    ;;
-esac
-
-unset __hookbook_shell
-unset __hookbook_shellname
+}
+
+[[ "${__hookbook_shellname}" == "zsh" ]] && __hookbook_setup_zsh && \unset __hookbook_shell __hookbook_shellname && \return 0
```
Comment (burke) [inline hookbook.sh]: Kind of too bad that this creates two more toplevel names. Could we instead do something like:

```
[[ "${__hookbook_shellname}" == "zsh" ]] && {

}
[[ "${__hookbook_shellname}" == "bash" ]] && {

}
\unset __hookbook_shell __hookbook_shellname
```

File: hookbook.sh [hunk 7]
```diff
@@ -108,36 +110,35 @@ case "${__hookbook_shellname}" in
...
+[[ "${__hookbook_shellname}" == "bash" ]] && __hookbook_setup_bash && \unset __hookbook_shell __hookbook_shellname && \return 0
+
+>&2 \echo "hookbook is not compatible with your shell (${__hookbook_shell})"
+\return 1
```
Comment (burke) [inline hookbook.sh:138]: I'm not totally convinced this won't blow stuff up from this context. Are you sure this works?

File: test [hunk 8]
```diff
@@ -5,42 +5,54 @@
 # /usr/local/bin/bash is bash 4+
 # /usr/local/bin/zsh is installed
 
-require('open3')
-require('pty')
+require_relative 'shell'
+
+REQUIRE_ESCAPES = [
```
Comment (burke) [inline test]: IMO a great example of where `%w()` is actually the right choice.

**Review by burke — COMMENTED** [review]
Looking great. A few minor things.

File: hookbook.sh (hunk 6 above)
Comment (lugray) [reply hookbook.sh]: I did this change before discovering that pattern.  Thanks for the catch.

File: hookbook.sh (hunk 7 above)
Comment (lugray) [reply hookbook.sh:138]: Nope.  But this is just moved code:
https://github.com/Shopify/hookbook/blob/6ba17bc61dd55d3b11da76257f151d72764878d2/hookbook.sh#L134-L137

Comment (burke) [reply hookbook.sh:138]: Oh, ok then, carry on.

**Pull request merged**